<html><head><title>bwa_aln.pl</title>
<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1" >
<link rel="stylesheet" title="blkbluw" type="text/css" href="_blkbluw.css" media="all" >
<link rel="alternate stylesheet" title="blkmagw" type="text/css" href="_blkmagw.css" media="all" >
<link rel="alternate stylesheet" title="blkcynw" type="text/css" href="_blkcynw.css" media="all" >
<link rel="alternate stylesheet" title="whtprpk" type="text/css" href="_whtprpk.css" media="all" >
<link rel="alternate stylesheet" title="whtnavk" type="text/css" href="_whtnavk.css" media="all" >
<link rel="alternate stylesheet" title="grygrnk" type="text/css" href="_grygrnk.css" media="all" >
<link rel="alternate stylesheet" title="whtgrng" type="text/css" href="_whtgrng.css" media="all" >
<link rel="alternate stylesheet" title="blkgrng" type="text/css" href="_blkgrng.css" media="all" >
<link rel="alternate stylesheet" title="grygrnw" type="text/css" href="_grygrnw.css" media="all" >
<link rel="alternate stylesheet" title="blkbluw" type="text/css" href="_blkbluw.css" media="all" >
<link rel="alternate stylesheet" title="whtpurk" type="text/css" href="_whtpurk.css" media="all" >
<link rel="alternate stylesheet" title="whtgrng" type="text/css" href="_whtgrng.css" media="all" >
<link rel="alternate stylesheet" title="grygrnw" type="text/css" href="_grygrnw.css" media="all" >

<script type="text/javascript" src="_podly.js"></script>

</head>
<body class='pod'>
<!--
  generated by Pod::Simple::HTML v3.28,
  using Pod::Simple::PullParser v3.28,
  under Perl v5.016003 at Wed Feb  5 16:23:50 2014 GMT.

 If you want to change this HTML document, you probably shouldn't do that
   by changing it directly.  Instead, see about changing the calling options
   to Pod::Simple::HTML, and/or subclassing Pod::Simple::HTML,
   then reconverting this document from the Pod source.
   When in doubt, email the author of Pod::Simple::HTML for advice.
   See 'perldoc Pod::Simple::HTML' for more info.

-->

<!-- start doc -->
<p class="backlinktop"><b><a name="___top" href="index.html" accesskey="1" title="All Documents">&lt;&lt;</a></b></p>

<div class='indexgroup'>
<ul   class='indexList indexList1'>
  <li class='indexItem indexItem1'><a href='#NAME'>NAME</a>
  <li class='indexItem indexItem1'><a href='#SYNOPSIS'>SYNOPSIS</a>
  <li class='indexItem indexItem1'><a href='#OPTIONS'>OPTIONS</a>
  <ul   class='indexList indexList2'>
    <li class='indexItem indexItem2'><a href='#TARGETED_PROCESSING'>TARGETED PROCESSING</a>
    <li class='indexItem indexItem2'><a href='#INPUT_FILE_TYPES'>INPUT FILE TYPES</a>
  </ul>
  <li class='indexItem indexItem1'><a href='#DESCRIPTION'>DESCRIPTION</a>
</ul>
</div>

<h1><a class='u' href='#___top' title='click to go to top of document'
name="NAME"
>NAME</a></h1>

<p>bwa_aln.pl - Align a set of lanes to specified reference with single command.</p>

<h1><a class='u' href='#___top' title='click to go to top of document'
name="SYNOPSIS"
>SYNOPSIS</a></h1>

<p>bwa_aln.pl [options] [file(s)...]</p>

<pre>  Required parameters:
    -outdir    -o   Folder to output result to.
    -reference -r   Path to reference genome file *.fa[.gz]
    -sample    -s   Sample name to be applied to output file.
    -threads   -t   Number of threads to use. [1]

  Targeted processing:
    -process   -p   Only process this step then exit, optionally set -index
                      bwamem - only applicable if input is bam
                        mark - Run duplicate marking (-index N/A)

    -index     -i   Optionally restrict &#39;-p&#39; to single job
                      bwamem - 1..&#60;lane_count&#62;

  Other:
    -help      -h   Brief help message.
    -man       -m   Full documentation.

  File list can be full file names or wildcard, e.g.
    bwa_aln.pl -t 16 -r some/genome.fa.gz -o myout -s sample input/*.fq.gz

  Run with &#39;-m&#39; for possible input file types.</pre>

<h1><a class='u' href='#___top' title='click to go to top of document'
name="OPTIONS"
>OPTIONS</a></h1>

<dl>
<dt><a name="-outdir"
><b>-outdir</b></a></dt>

<dd>
<p>Directory to write output to. During processing a temp folder will be generated in this area, should the process fail <b>only delete this if</b> you are unable to resume the process.</p>

<p>Final output files include: &#60;SAMPLE&#62;.bam, &#60;SAMPLE&#62;.bam.bai, &#60;SAMPLE&#62;.md5, &#60;SAMPLE&#62;.met</p>

<dt><a name="-reference"
><b>-reference</b></a></dt>

<dd>
<p>Path to genome.fa[.gz] file and associated indexes for BWA.</p>

<dt><a name="-sample"
><b>-sample</b></a></dt>

<dd>
<p>Name to be applied to output files. Special characters will not be magically fixed.</p>

<dt><a name="-threads"
><b>-threads</b></a></dt>

<dd>
<p>Number of threads to be used in processing.</p>

<p>If perl is not compiled with threading some steps will not run in parallel, however much of the script calls other tools that will still utilise this appropriately.</p>

<dt><a name="-help"
><b>-help</b></a></dt>

<dd>
<p>Print a brief help message and exits.</p>

<dt><a name="-man"
><b>-man</b></a></dt>

<dd>
<p>Prints the manual page and exits.</p>
</dd>
</dl>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="TARGETED_PROCESSING"
>TARGETED PROCESSING</a></h2>

<dl>
<dt><a name="-process"
><b>-process</b></a></dt>

<dd>
<p>If you want to run the code in a more efficient manner then this allows each procesing type to be executed in isolation. You can restrict to a single process within the block by specifying <b>-index</b> as well.</p>
</dd>
</dl>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="INPUT_FILE_TYPES"
>INPUT FILE TYPES</a></h2>

<p>There are several types of file that the script is able to process.</p>

<dl>
<dt><a name="f[ast]q"
>f[ast]q</a></dt>

<dd>
<p>A standard uncompressed fastq file. Requires a pair of inputs with standard suffix of &#39;_1&#39; and &#39;_2&#39; immediately prior to &#39;.f[ast]q&#39;.</p>

<dt><a name="f[ast]q.gz"
>f[ast]q.gz</a></dt>

<dd>
<p>As *.f[ast]q but compressed with gzip.</p>

<dt><a name="bam"
>bam</a></dt>

<dd>
<p>A list of single lane BAM files, no information is taken from the headers.</p>

<p><b>This method has additional processing converted to *.fq.gz to give common start point.</b></p>
</dd>
</dl>

<h1><a class='u' href='#___top' title='click to go to top of document'
name="DESCRIPTION"
>DESCRIPTION</a></h1>

<p><b>bwa_aln.pl</b> will attempt to run all mapping steps for BWA ALN and subsequent duplicate marking automatically.</p>
<p class="backlinkbottom"><b><a name="___bottom" href="index.html" title="All Documents">&lt;&lt;</a></b></p>

<!-- end doc -->

</body></html>
